lib/utils.nix

```
@@ -125,27 +125,185 @@ rec {
+    }
+    nix-repl> buildSurround 1
+    {
+      close = ")";
+      open = "(";
+    }
+    ```
+
+    We invoke `buildSurround` builder starting a depth 0, and increasing the
+    depth until we have a Surround where `surround.close` is not in the
+    string we're escaping.
+
+    We return the original string, first (possibly, depending on
+    `surroundWithNewlines`) surrounded by newlines , and secondly surrounded by
+    the Surround we found above.
+    ```
+  */
+  escapeString =
+    {
+      string,
+      buildSurround,
+      surroundWithNewlines,
+    }:
+    let
+      findSurround =
+        {
+          string,
+          buildSurround,
+          depth,
+        }:
+        let
+          surround = buildSurround depth;
+        in
+        if lib.hasInfix surround.close string then
+          findSurround {
+            inherit string buildSurround;
+            depth = depth + 1;
+          }
+        else
+          surround;
+
+      surround = findSurround {
+        inherit string buildSurround;
+        depth = 0;
+      };
+      newlines = if surroundWithNewlines then "\n" else "";
+    in
+    "${surround.open}${newlines}${string}${newlines}${surround.close}";
+
+  /**
+    Given a string, escape it as a Lua [long literal].
+
+    [long literal]: https://www.lua.org/manual/5.4/manual.html#3.1
+
+    # Examples
+
+    ```nix
+    nix-repl> escapeStringForLua "simple"
+    "[[simple]]"
+    ```
+
+    ```nix
+    nix-repl> escapeStringForLua "]]"
+    "[=[]]]=]"
+    ```
+
+    # Type
+
+    ```
+    escapeStringForLua :: String -> String
+    ```
+  */
+  escapeStringForLua =
+    string:
+    escapeString {
+      inherit string;
+      surroundWithNewlines = false;
+
+      # [[
+      # [=[
+      # [==[
+      # ...
+      buildSurround =
+        depth:
+        let
+          equals = lib.strings.replicate depth "=";
+        in
+        {
+          open = "[${equals}[";
+          close = "]${equals}]";
+        };
+    };
+
+  /**
+    Given a string, escape it as a Vimscript [:let-heredoc], for use by [:lua].
+
+    [:let-heredoc]: https://neovim.io/doc/user/eval.html#%3Alet-heredoc
+    [:lua]: https://neovim.io/doc/user/lua.html#%3Alua-heredoc
+
+    # Examples
+
+    ```nix
+    escapeStringForVimscript "simple"
+    => "<< EOF\nsimple\nEOF"
+    ```
+
+    ```nix
+    escapeStringForVimscript "EOF"
+    => "<< EOFF\nEOF\nEOFF"
+    ```
+
+    # Type
+
+    ```
+    escapeStringForVimscript :: String -> String
+    ```
+  */
+  escapeStringForVimscript =
+    string:
+    (
+      let
+        escapedString = escapeString {
+          inherit string;
+          surroundWithNewlines = true;
+
+          # EOF
+          # EOFF
+          # EOFFF
+          # ...
+          buildSurround =
+            depth:
+            let
+              fs = lib.strings.replicate (depth + 1) "F";
+              token = "EO${fs}";
+            in
+            {
+              open = token;
```

Would including the `<< ` here make the usage below simpler? There'd no longer be a need to do `"<< ${escapedString}"`...


```suggestion
              open = "<< ${token}";
```